# Flight Price Prediction Feature Engineering Advance Level

## Importing importand modules

In [103]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

## Importing Datasets

In [104]:
train=pd.read_excel(r"C:\Users\ashee\OneDrive\Desktop\Krish Black Frida Feature Engineering\Data_Train.xlsx")

In [105]:
train

Airline Date_of_Journey    Source Destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

In [106]:
test=pd.read_excel(r"C:\Users\ashee\OneDrive\Desktop\Krish Black Frida Feature Engineering\Test_set.xlsx")

In [107]:
test

Airline Date_of_Journey    Source Destination  \
0           Jet Airways       6/06/2019     Delhi      Cochin   
1                IndiGo      12/05/2019   Kolkata    Banglore   
2           Jet Airways      21/05/2019     Delhi      Cochin   
3     Multiple carriers      21/05/2019     Delhi      Cochin   
4              Air Asia      24/06/2019  Banglore       Delhi   
...                 ...             ...       ...         ...   
2666          Air India       6/06/2019   Kolkata    Banglore   
2667             IndiGo      27/03/2019   Kolkata    Banglore   
2668        Jet Airways       6/03/2019     Delhi      Cochin   
2669          Air India       6/03/2019     Delhi      Cochin   
2670  Multiple carriers      15/06/2019     Delhi      Cochin   

                Route Dep_Time  Arrival_Time Duration Total_Stops  \
0     DEL → BOM → COK    17:30  04:25 07 Jun  10h 55m      1 stop   
1     CCU → MAA → BLR    06:20         10:20       4h      1 stop   
2     DEL → BOM → COK    19:15  19:00 22 May  23h 45m      1 stop   
3     DEL → BOM → COK    08:00         21:00      13h      1 stop   
4           BLR → DEL    23:55  02:45 25 Jun   2h 50m    non-stop   
...               ...      ...           ...      ...         ...   
2666  CCU → DEL → BLR    20:30  20:25 07 Jun  23h 55m      1 stop   
2667        CCU → BLR    14:20         16:55   2h 35m    non-stop   
2668  DEL → BOM → COK    21:50  04:25 07 Mar   6h 35m      1 stop   
2669  DEL → BOM → COK    04:00         19:15  15h 15m      1 stop   
2670  DEL → BOM → COK    04:55         19:15  14h 20m      1 stop   

                  Additional_Info  
0                         No info  
1                         No info  
2     In-flight meal not included  
3                         No info  
4                         No info  
...                           ...  
2666                      No info  
2667                      No info  
2668                      No info  
2669                      No info  
2670                      No info  

[2671 rows x 10 columns]

#### Note : There are two dataset we get from kaggle train and test we need to combine them then only we can do coding on that.

In [108]:
dataset=pd.concat([train,test],axis=0)

In [109]:
dataset

Airline Date_of_Journey    Source Destination  \
0                IndiGo      24/03/2019  Banglore   New Delhi   
1             Air India       1/05/2019   Kolkata    Banglore   
2           Jet Airways       9/06/2019     Delhi      Cochin   
3                IndiGo      12/05/2019   Kolkata    Banglore   
4                IndiGo      01/03/2019  Banglore   New Delhi   
...                 ...             ...       ...         ...   
2666          Air India       6/06/2019   Kolkata    Banglore   
2667             IndiGo      27/03/2019   Kolkata    Banglore   
2668        Jet Airways       6/03/2019     Delhi      Cochin   
2669          Air India       6/03/2019     Delhi      Cochin   
2670  Multiple carriers      15/06/2019     Delhi      Cochin   

                      Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                 BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1     CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2     DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3           CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4           BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                     ...      ...           ...      ...         ...   
2666        CCU → DEL → BLR    20:30  20:25 07 Jun  23h 55m      1 stop   
2667              CCU → BLR    14:20         16:55   2h 35m    non-stop   
2668        DEL → BOM → COK    21:50  04:25 07 Mar   6h 35m      1 stop   
2669        DEL → BOM → COK    04:00         19:15  15h 15m      1 stop   
2670        DEL → BOM → COK    04:55         19:15  14h 20m      1 stop   

     Additional_Info    Price  
0            No info   3897.0  
1            No info   7662.0  
2            No info  13882.0  
3            No info   6218.0  
4            No info  13302.0  
...              ...      ...  
2666         No info      NaN  
2667         No info      NaN  
2668         No info      NaN  
2669         No info      NaN  
2670         No info      NaN  

[13354 rows x 11 columns]

In [110]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13354 entries, 0 to 2670
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13354 non-null  object 
 1   Date_of_Journey  13354 non-null  object 
 2   Source           13354 non-null  object 
 3   Destination      13354 non-null  object 
 4   Route            13353 non-null  object 
 5   Dep_Time         13354 non-null  object 
 6   Arrival_Time     13354 non-null  object 
 7   Duration         13354 non-null  object 
 8   Total_Stops      13353 non-null  object 
 9   Additional_Info  13354 non-null  object 
 10  Price            10683 non-null  float64
dtypes: float64(1), object(10)
memory usage: 1.2+ MB


In [111]:
dataset.describe()

Price
count  10683.000000
mean    9087.064121
std     4611.359167
min     1759.000000
25%     5277.000000
50%     8372.000000
75%    12373.000000
max    79512.000000

In [112]:
dataset.isnull().sum()

Airline               0
Date_of_Journey       0
Source                0
Destination           0
Route                 1
Dep_Time              0
Arrival_Time          0
Duration              0
Total_Stops           1
Additional_Info       0
Price              2671
dtype: int64

In [113]:
[feature for feature in dataset.columns if dataset[feature].isnull().sum()>=1]

['Route', 'Total_Stops', 'Price']

### Observation: 
1. The dataset has only 2 null values (1 in Route and 1 in Total_Stops).
2. The Price column null value of 2671 is test null for model training.
3. The price column is numerical data; the rest is categorical data.

In [114]:
dataset.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info    Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897.0  
1    05:50         13:15   7h 25m     2 stops         No info   7662.0  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882.0  
3    18:05         23:30   5h 25m      1 stop         No info   6218.0  
4    16:50         21:35   4h 45m      1 stop         No info  13302.0

In [115]:
dataset['Date_of_Journey'].str.split('/').str[0]

0       24
1        1
2        9
3       12
4       01
        ..
2666     6
2667    27
2668     6
2669     6
2670    15
Name: Date_of_Journey, Length: 13354, dtype: object

In [116]:
dataset['Day']=dataset['Date_of_Journey'].str.split('/').str[0]
dataset['Month']=dataset['Date_of_Journey'].str.split('/').str[1]
dataset['Year']=dataset['Date_of_Journey'].str.split('/').str[2]

In [117]:
dataset

Airline Date_of_Journey    Source Destination  \
0                IndiGo      24/03/2019  Banglore   New Delhi   
1             Air India       1/05/2019   Kolkata    Banglore   
2           Jet Airways       9/06/2019     Delhi      Cochin   
3                IndiGo      12/05/2019   Kolkata    Banglore   
4                IndiGo      01/03/2019  Banglore   New Delhi   
...                 ...             ...       ...         ...   
2666          Air India       6/06/2019   Kolkata    Banglore   
2667             IndiGo      27/03/2019   Kolkata    Banglore   
2668        Jet Airways       6/03/2019     Delhi      Cochin   
2669          Air India       6/03/2019     Delhi      Cochin   
2670  Multiple carriers      15/06/2019     Delhi      Cochin   

                      Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                 BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1     CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2     DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3           CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4           BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                     ...      ...           ...      ...         ...   
2666        CCU → DEL → BLR    20:30  20:25 07 Jun  23h 55m      1 stop   
2667              CCU → BLR    14:20         16:55   2h 35m    non-stop   
2668        DEL → BOM → COK    21:50  04:25 07 Mar   6h 35m      1 stop   
2669        DEL → BOM → COK    04:00         19:15  15h 15m      1 stop   
2670        DEL → BOM → COK    04:55         19:15  14h 20m      1 stop   

     Additional_Info    Price Day Month  Year  
0            No info   3897.0  24    03  2019  
1            No info   7662.0   1    05  2019  
2            No info  13882.0   9    06  2019  
3            No info   6218.0  12    05  2019  
4            No info  13302.0  01    03  2019  
...              ...      ...  ..   ...   ...  
2666         No info      NaN   6    06  2019  
2667         No info      NaN  27    03  2019  
2668         No info      NaN   6    03  2019  
2669         No info      NaN   6    03  2019  
2670         No info      NaN  15    06  2019  

[13354 rows x 14 columns]

In [118]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13354 entries, 0 to 2670
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13354 non-null  object 
 1   Date_of_Journey  13354 non-null  object 
 2   Source           13354 non-null  object 
 3   Destination      13354 non-null  object 
 4   Route            13353 non-null  object 
 5   Dep_Time         13354 non-null  object 
 6   Arrival_Time     13354 non-null  object 
 7   Duration         13354 non-null  object 
 8   Total_Stops      13353 non-null  object 
 9   Additional_Info  13354 non-null  object 
 10  Price            10683 non-null  float64
 11  Day              13354 non-null  object 
 12  Month            13354 non-null  object 
 13  Year             13354 non-null  object 
dtypes: float64(1), object(13)
memory usage: 1.5+ MB


In [119]:
dataset['Day']=dataset['Day'].astype(int)
dataset['Month']=dataset['Month'].astype(int)
dataset['Year']=dataset['Year'].astype(int)

In [120]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13354 entries, 0 to 2670
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13354 non-null  object 
 1   Date_of_Journey  13354 non-null  object 
 2   Source           13354 non-null  object 
 3   Destination      13354 non-null  object 
 4   Route            13353 non-null  object 
 5   Dep_Time         13354 non-null  object 
 6   Arrival_Time     13354 non-null  object 
 7   Duration         13354 non-null  object 
 8   Total_Stops      13353 non-null  object 
 9   Additional_Info  13354 non-null  object 
 10  Price            10683 non-null  float64
 11  Day              13354 non-null  int32  
 12  Month            13354 non-null  int32  
 13  Year             13354 non-null  int32  
dtypes: float64(1), int32(3), object(10)
memory usage: 1.4+ MB


In [121]:
dataset.drop('Date_of_Journey',axis=1,inplace=True)

In [122]:
dataset.head()

Airline    Source Destination                  Route Dep_Time  \
0       IndiGo  Banglore   New Delhi              BLR → DEL    22:20   
1    Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR    05:50   
2  Jet Airways     Delhi      Cochin  DEL → LKO → BOM → COK    09:25   
3       IndiGo   Kolkata    Banglore        CCU → NAG → BLR    18:05   
4       IndiGo  Banglore   New Delhi        BLR → NAG → DEL    16:50   

   Arrival_Time Duration Total_Stops Additional_Info    Price  Day  Month  \
0  01:10 22 Mar   2h 50m    non-stop         No info   3897.0   24      3   
1         13:15   7h 25m     2 stops         No info   7662.0    1      5   
2  04:25 10 Jun      19h     2 stops         No info  13882.0    9      6   
3         23:30   5h 25m      1 stop         No info   6218.0   12      5   
4         21:35   4h 45m      1 stop         No info  13302.0    1      3   

   Year  
0  2019  
1  2019  
2  2019  
3  2019  
4  2019

### Obserbation :
1. There was Date_of_Journey in the biggining so we splited  Day,Month,Year  in three parts for machine learning
2. Day,Month and Year are string formate ,converted in to int fromat.

In [123]:
dataset['Arrival_Time']=dataset['Arrival_Time'].str.split(' ').str[0]

In [124]:
dataset.head(1)

Airline    Source Destination      Route Dep_Time Arrival_Time Duration  \
0  IndiGo  Banglore   New Delhi  BLR → DEL    22:20        01:10   2h 50m   

  Total_Stops Additional_Info   Price  Day  Month  Year  
0    non-stop         No info  3897.0   24      3  2019

In [125]:
dataset['Arrival_Hours']=dataset['Arrival_Time'].str.split(':').str[0]
dataset['Arrival_Minutes']=dataset['Arrival_Time'].str.split(':').str[1]

In [126]:
dataset.head()

Airline    Source Destination                  Route Dep_Time  \
0       IndiGo  Banglore   New Delhi              BLR → DEL    22:20   
1    Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR    05:50   
2  Jet Airways     Delhi      Cochin  DEL → LKO → BOM → COK    09:25   
3       IndiGo   Kolkata    Banglore        CCU → NAG → BLR    18:05   
4       IndiGo  Banglore   New Delhi        BLR → NAG → DEL    16:50   

  Arrival_Time Duration Total_Stops Additional_Info    Price  Day  Month  \
0        01:10   2h 50m    non-stop         No info   3897.0   24      3   
1        13:15   7h 25m     2 stops         No info   7662.0    1      5   
2        04:25      19h     2 stops         No info  13882.0    9      6   
3        23:30   5h 25m      1 stop         No info   6218.0   12      5   
4        21:35   4h 45m      1 stop         No info  13302.0    1      3   

   Year Arrival_Hours Arrival_Minutes  
0  2019            01              10  
1  2019            13              15  
2  2019            04              25  
3  2019            23              30  
4  2019            21              35

In [127]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13354 entries, 0 to 2670
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13354 non-null  object 
 1   Source           13354 non-null  object 
 2   Destination      13354 non-null  object 
 3   Route            13353 non-null  object 
 4   Dep_Time         13354 non-null  object 
 5   Arrival_Time     13354 non-null  object 
 6   Duration         13354 non-null  object 
 7   Total_Stops      13353 non-null  object 
 8   Additional_Info  13354 non-null  object 
 9   Price            10683 non-null  float64
 10  Day              13354 non-null  int32  
 11  Month            13354 non-null  int32  
 12  Year             13354 non-null  int32  
 13  Arrival_Hours    13354 non-null  object 
 14  Arrival_Minutes  13354 non-null  object 
dtypes: float64(1), int32(3), object(11)
memory usage: 1.5+ MB


In [128]:
dataset.drop('Arrival_Time',axis=1,inplace=True)

In [129]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13354 entries, 0 to 2670
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13354 non-null  object 
 1   Source           13354 non-null  object 
 2   Destination      13354 non-null  object 
 3   Route            13353 non-null  object 
 4   Dep_Time         13354 non-null  object 
 5   Duration         13354 non-null  object 
 6   Total_Stops      13353 non-null  object 
 7   Additional_Info  13354 non-null  object 
 8   Price            10683 non-null  float64
 9   Day              13354 non-null  int32  
 10  Month            13354 non-null  int32  
 11  Year             13354 non-null  int32  
 12  Arrival_Hours    13354 non-null  object 
 13  Arrival_Minutes  13354 non-null  object 
dtypes: float64(1), int32(3), object(10)
memory usage: 1.4+ MB


In [130]:
dataset['Arrival_Hours']=dataset['Arrival_Hours'].astype(int)
dataset['Arrival_Minutes']=dataset['Arrival_Minutes'].astype(int)

In [131]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13354 entries, 0 to 2670
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13354 non-null  object 
 1   Source           13354 non-null  object 
 2   Destination      13354 non-null  object 
 3   Route            13353 non-null  object 
 4   Dep_Time         13354 non-null  object 
 5   Duration         13354 non-null  object 
 6   Total_Stops      13353 non-null  object 
 7   Additional_Info  13354 non-null  object 
 8   Price            10683 non-null  float64
 9   Day              13354 non-null  int32  
 10  Month            13354 non-null  int32  
 11  Year             13354 non-null  int32  
 12  Arrival_Hours    13354 non-null  int32  
 13  Arrival_Minutes  13354 non-null  int32  
dtypes: float64(1), int32(5), object(8)
memory usage: 1.3+ MB


In [132]:
dataset['Dep_Hour']=dataset['Dep_Time'].str.split(':').str[0]
dataset['Dep_Minutes']=dataset['Dep_Time'].str.split(':').str[1]

In [133]:
dataset['Dep_Hour']=dataset['Dep_Hour'].astype(int)
dataset['Dep_Minutes']=dataset['Dep_Minutes'].astype(int)

In [134]:
dataset.drop('Dep_Time',axis=1,inplace=True)

In [135]:
dataset.head()

Airline    Source Destination                  Route Duration  \
0       IndiGo  Banglore   New Delhi              BLR → DEL   2h 50m   
1    Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR   7h 25m   
2  Jet Airways     Delhi      Cochin  DEL → LKO → BOM → COK      19h   
3       IndiGo   Kolkata    Banglore        CCU → NAG → BLR   5h 25m   
4       IndiGo  Banglore   New Delhi        BLR → NAG → DEL   4h 45m   

  Total_Stops Additional_Info    Price  Day  Month  Year  Arrival_Hours  \
0    non-stop         No info   3897.0   24      3  2019              1   
1     2 stops         No info   7662.0    1      5  2019             13   
2     2 stops         No info  13882.0    9      6  2019              4   
3      1 stop         No info   6218.0   12      5  2019             23   
4      1 stop         No info  13302.0    1      3  2019             21   

   Arrival_Minutes  Dep_Hour  Dep_Minutes  
0               10        22           20  
1               15         5           50  
2               25         9           25  
3               30        18            5  
4               35        16           50

### Obserbation :
We splited two columns Arrival_time and Dep_time into hours and minutes and both converted string to int format

In [136]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13354 entries, 0 to 2670
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13354 non-null  object 
 1   Source           13354 non-null  object 
 2   Destination      13354 non-null  object 
 3   Route            13353 non-null  object 
 4   Duration         13354 non-null  object 
 5   Total_Stops      13353 non-null  object 
 6   Additional_Info  13354 non-null  object 
 7   Price            10683 non-null  float64
 8   Day              13354 non-null  int32  
 9   Month            13354 non-null  int32  
 10  Year             13354 non-null  int32  
 11  Arrival_Hours    13354 non-null  int32  
 12  Arrival_Minutes  13354 non-null  int32  
 13  Dep_Hour         13354 non-null  int32  
 14  Dep_Minutes      13354 non-null  int32  
dtypes: float64(1), int32(7), object(7)
memory usage: 1.3+ MB


In [137]:
dataset['Total_Stops'].unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

In [138]:
dataset[dataset['Total_Stops'].isnull()]

Airline Source Destination Route Duration Total_Stops Additional_Info  \
9039  Air India  Delhi      Cochin   NaN  23h 40m         NaN         No info   

       Price  Day  Month  Year  Arrival_Hours  Arrival_Minutes  Dep_Hour  \
9039  7480.0    6      5  2019              9               25         9   

      Dep_Minutes  
9039           45

In [139]:
dataset.drop(9039,axis=0,inplace=True)

In [140]:
dataset['Total_Stops'].unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', '4 stops'],
      dtype=object)

In [141]:
dataset['Total_Stops']=dataset['Total_Stops'].map({'non-stop':0, '2 stops':2, '1 stop':1, '3 stops':3, '4 stops':4})

### Obserbation: 
We changed Total_Stops columns into numbers and there where one null values we deleted that also

In [142]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13353 entries, 0 to 2670
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13353 non-null  object 
 1   Source           13353 non-null  object 
 2   Destination      13353 non-null  object 
 3   Route            13353 non-null  object 
 4   Duration         13353 non-null  object 
 5   Total_Stops      13353 non-null  int64  
 6   Additional_Info  13353 non-null  object 
 7   Price            10682 non-null  float64
 8   Day              13353 non-null  int32  
 9   Month            13353 non-null  int32  
 10  Year             13353 non-null  int32  
 11  Arrival_Hours    13353 non-null  int32  
 12  Arrival_Minutes  13353 non-null  int32  
 13  Dep_Hour         13353 non-null  int32  
 14  Dep_Minutes      13353 non-null  int32  
dtypes: float64(1), int32(7), int64(1), object(6)
memory usage: 1.3+ MB


In [143]:
dataset['Additional_Info'] = dataset['Additional_Info'].replace('No info', 'No Info')


In [144]:
dataset['Additional_Info'].unique()

array(['No Info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover',
       '1 Long layover', 'Change airports', 'Business class',
       'Red-eye flight', '2 Long layover'], dtype=object)

In [145]:
dataset.head()

Airline    Source Destination                  Route Duration  \
0       IndiGo  Banglore   New Delhi              BLR → DEL   2h 50m   
1    Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR   7h 25m   
2  Jet Airways     Delhi      Cochin  DEL → LKO → BOM → COK      19h   
3       IndiGo   Kolkata    Banglore        CCU → NAG → BLR   5h 25m   
4       IndiGo  Banglore   New Delhi        BLR → NAG → DEL   4h 45m   

   Total_Stops Additional_Info    Price  Day  Month  Year  Arrival_Hours  \
0            0         No Info   3897.0   24      3  2019              1   
1            2         No Info   7662.0    1      5  2019             13   
2            2         No Info  13882.0    9      6  2019              4   
3            1         No Info   6218.0   12      5  2019             23   
4            1         No Info  13302.0    1      3  2019             21   

   Arrival_Minutes  Dep_Hour  Dep_Minutes  
0               10        22           20  
1               15         5           50  
2               25         9           25  
3               30        18            5  
4               35        16           50

In [146]:
dataset['Duration_Hour']=dataset['Duration'].str.split(' ').str[0].str.split('h').str[0]

In [147]:
dataset['Duration_Minutes']=dataset['Duration'].str.split(' ').str[1].str.split('m').str[0]

In [150]:
dataset[dataset['Duration_Hour']=='5m']

Airline  Source Destination                  Route Duration  \
6474  Air India  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD       5m   
2660  Air India  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD       5m   

      Total_Stops Additional_Info    Price  Day  Month  Year  Arrival_Hours  \
6474            2         No Info  17327.0    6      3  2019             16   
2660            2         No Info      NaN   12      3  2019             16   

      Arrival_Minutes  Dep_Hour  Dep_Minutes Duration_Hour Duration_Minutes  
6474               55        16           50            5m              NaN  
2660               55        16           50            5m              NaN

In [151]:
dataset.drop([6474,2660],axis=0,inplace=True)

In [152]:
dataset['Duration_Hour']=dataset['Duration_Hour'].astype(int)

In [153]:
dataset[dataset['Duration_Minutes'].isnull()]

Airline    Source Destination                  Route Duration  \
2           Jet Airways     Delhi      Cochin  DEL → LKO → BOM → COK      19h   
18            Air India     Delhi      Cochin  DEL → CCU → BOM → COK      23h   
33          Jet Airways     Delhi      Cochin  DEL → NAG → BOM → COK      22h   
44    Multiple carriers     Delhi      Cochin        DEL → BOM → COK      12h   
53               IndiGo  Banglore       Delhi              BLR → DEL       3h   
...                 ...       ...         ...                    ...      ...   
2588          Air India     Delhi      Cochin              DEL → COK       3h   
2598        Jet Airways     Delhi      Cochin        DEL → BOM → COK      11h   
2604  Multiple carriers     Delhi      Cochin        DEL → BOM → COK      10h   
2607  Multiple carriers     Delhi      Cochin        DEL → BOM → COK      13h   
2622        Jet Airways  Banglore       Delhi              BLR → DEL       3h   

      Total_Stops              Additional_Info    Price  Day  Month  Year  \
2               2                      No Info  13882.0    9      6  2019   
18              2                      No Info  13381.0   12      6  2019   
33              2  In-flight meal not included  10919.0   15      6  2019   
44              1                      No Info  13062.0   21      3  2019   
53              0                      No Info   3943.0   18      6  2019   
...           ...                          ...      ...  ...    ...   ...   
2588            0                      No Info      NaN   27      3  2019   
2598            1                      No Info      NaN    9      6  2019   
2604            1                      No Info      NaN   12      6  2019   
2607            1                      No Info      NaN    3      3  2019   
2622            0  In-flight meal not included      NaN    3      5  2019   

      Arrival_Hours  Arrival_Minutes  Dep_Hour  Dep_Minutes  Duration_Hour  \
2                 4               25         9           25             19   
18               19               15        20           15             23   
33               12               35        14           35             22   
44               21                0         9            0             12   
53                0               15        21           15              3   
...             ...              ...       ...          ...            ...   
2588             17               10        14           10              3   
2598             19                0         8            0             11   
2604             19               15         9           15             10   
2607             21                0         8            0             13   
2622             20               45        17           45              3   

     Duration_Minutes  
2                 NaN  
18                NaN  
33                NaN  
44                NaN  
53                NaN  
...               ...  
2588              NaN  
2598              NaN  
2604              NaN  
2607              NaN  
2622              NaN  

[1283 rows x 17 columns]

In [157]:
dataset['Duration_Minutes']=dataset['Duration_Minutes'].astype(int)

In [162]:
dataset['Duration_Hour']=dataset['Duration_Hour']*60

In [164]:
dataset['total_duration_minutes']=dataset['Duration_Hour']+dataset['Duration_Minutes']

In [166]:
dataset.head()

Airline    Source Destination                  Route Duration  \
0       IndiGo  Banglore   New Delhi              BLR → DEL   2h 50m   
1    Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR   7h 25m   
2  Jet Airways     Delhi      Cochin  DEL → LKO → BOM → COK      19h   
3       IndiGo   Kolkata    Banglore        CCU → NAG → BLR   5h 25m   
4       IndiGo  Banglore   New Delhi        BLR → NAG → DEL   4h 45m   

   Total_Stops Additional_Info    Price  Day  Month  Year  Arrival_Hours  \
0            0         No Info   3897.0   24      3  2019              1   
1            2         No Info   7662.0    1      5  2019             13   
2            2         No Info  13882.0    9      6  2019              4   
3            1         No Info   6218.0   12      5  2019             23   
4            1         No Info  13302.0    1      3  2019             21   

   Arrival_Minutes  Dep_Hour  Dep_Minutes  Duration_Hour  Duration_Minutes  \
0               10        22           20            120                50   
1               15         5           50            420                25   
2               25         9           25           1140                 0   
3               30        18            5            300                25   
4               35        16           50            240                45   

   total_duration_minutes  
0                     170  
1                     445  
2                    1140  
3                     325  
4                     285

In [168]:
dataset.drop(['Duration','Duration_Hour','Duration_Minutes'],axis=1,inplace=True)

In [169]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13350 entries, 0 to 2670
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Airline                 13350 non-null  object 
 1   Source                  13350 non-null  object 
 2   Destination             13350 non-null  object 
 3   Route                   13350 non-null  object 
 4   Total_Stops             13350 non-null  int64  
 5   Additional_Info         13350 non-null  object 
 6   Price                   10680 non-null  float64
 7   Day                     13350 non-null  int32  
 8   Month                   13350 non-null  int32  
 9   Year                    13350 non-null  int32  
 10  Arrival_Hours           13350 non-null  int32  
 11  Arrival_Minutes         13350 non-null  int32  
 12  Dep_Hour                13350 non-null  int32  
 13  Dep_Minutes             13350 non-null  int32  
 14  total_duration_minutes  13350 non-null  int3

### Obserbation
1. I splited 'Duration' columns into Duration_hours and duration_minutes.
2. There where 5m wrong data in duration hours i drop that.
3. There where null values in duration_minutes i just fill by 0.
4. Duration_hours and Duration_minutes both there string , converted into int format
5. We just added duration_hours(converted in minutes) and duration minutes,made total_duration_minutes.

### Note: We will do label encoding for other columns like Airline,Source,Destination,Route and Additional_ Info

In [170]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()

In [172]:
dataset['Airline']=le.fit_transform(dataset['Airline'])
dataset['Source']=le.fit_transform(dataset['Source'])
dataset['Destination']=le.fit_transform(dataset['Destination'])
dataset['Route']=le.fit_transform(dataset['Route'])
dataset['Additional_Info']=le.fit_transform(dataset['Additional_Info'])

In [173]:
dataset.head()

Airline  Source  Destination  Route  Total_Stops  Additional_Info    Price  \
0        3       0            5     18            0                6   3897.0   
1        1       3            0     87            2                6   7662.0   
2        4       2            1    122            2                6  13882.0   
3        3       3            0     95            1                6   6218.0   
4        3       0            5     29            1                6  13302.0   

   Day  Month  Year  Arrival_Hours  Arrival_Minutes  Dep_Hour  Dep_Minutes  \
0   24      3  2019              1               10        22           20   
1    1      5  2019             13               15         5           50   
2    9      6  2019              4               25         9           25   
3   12      5  2019             23               30        18            5   
4    1      3  2019             21               35        16           50   

   total_duration_minutes  
0                     170  
1                     445  
2                    1140  
3                     325  
4                     285

In [179]:
from sklearn.preprocessing import OneHotEncoder
ohe=OneHotEncoder(handle_unknown="ignore")

dataset['Airline']=ohe.fit_transform(dataset['Airline'])
dataset['Source']=ohe.fit_transform(dataset['Source'])
dataset['Destination']=ohe.fit_transform(dataset['Destination'])
dataset['Route']=ohe.fit_transform(dataset['Route'])
dataset['Additional_Info']=ohe.fit_transform(dataset['Additional_Info'])

dataset=pd.get_dummies(dataset,columns=['Airline','Source','Destination','Route','Additional_Info'],drop_first=True)

In [187]:
dataset.head()

Total_Stops    Price  Day  Month  Year  Arrival_Hours  Arrival_Minutes  \
0            0   3897.0   24      3  2019              1               10   
1            2   7662.0    1      5  2019             13               15   
2            2  13882.0    9      6  2019              4               25   
3            1   6218.0   12      5  2019             23               30   
4            1  13302.0    1      3  2019             21               35   

   Dep_Hour  Dep_Minutes  total_duration_minutes  ...  Route_131  \
0        22           20                     170  ...      False   
1         5           50                     445  ...      False   
2         9           25                    1140  ...      False   
3        18            5                     325  ...      False   
4        16           50                     285  ...      False   

   Additional_Info_0  Additional_Info_1  Additional_Info_2  Additional_Info_3  \
0              False              False              False              False   
1              False              False              False              False   
2              False              False              False              False   
3              False              False              False              False   
4              False              False              False              False   

   Additional_Info_4  Additional_Info_5  Additional_Info_6  Additional_Info_7  \
0              False              False               True              False   
1              False              False               True              False   
2              False              False               True              False   
3              False              False               True              False   
4              False              False               True              False   

   Additional_Info_8  
0              False  
1              False  
2              False  
3              False  
4              False  

[5 rows x 174 columns]

In [195]:
bool_cols = dataset.select_dtypes(include='bool').columns
dataset[bool_cols] = dataset[bool_cols].astype(int)


In [196]:
dataset.head()

Total_Stops    Price  Day  Month  Year  Arrival_Hours  Arrival_Minutes  \
0            0   3897.0   24      3  2019              1               10   
1            2   7662.0    1      5  2019             13               15   
2            2  13882.0    9      6  2019              4               25   
3            1   6218.0   12      5  2019             23               30   
4            1  13302.0    1      3  2019             21               35   

   Dep_Hour  Dep_Minutes  total_duration_minutes  ...  Route_131  \
0        22           20                     170  ...          0   
1         5           50                     445  ...          0   
2         9           25                    1140  ...          0   
3        18            5                     325  ...          0   
4        16           50                     285  ...          0   

   Additional_Info_0  Additional_Info_1  Additional_Info_2  Additional_Info_3  \
0                  0                  0                  0                  0   
1                  0                  0                  0                  0   
2                  0                  0                  0                  0   
3                  0                  0                  0                  0   
4                  0                  0                  0                  0   

   Additional_Info_4  Additional_Info_5  Additional_Info_6  Additional_Info_7  \
0                  0                  0                  1                  0   
1                  0                  0                  1                  0   
2                  0                  0                  1                  0   
3                  0                  0                  1                  0   
4                  0                  0                  1                  0   

   Additional_Info_8  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0  

[5 rows x 174 columns]

### Obserbation:
1. first of all we did LabelEncoding with categorical data like Airline,Source,Destination,Route and Additional_info  because we if directly do one hot 
encoding then str will not convert into 0 and 1 
2. we can do one hot encoding or pd.get_dummies , but OneHotEncoding is always recomended because when we make pickle file to deplyment then it needs
3. After doing one hot encoding we were getting boolian values, this is why we converted into 0 and 1 format 